### ISSUE TITLE
Optional chaining of unknown should be allowed

### ISSUE THREAD

**Issue by hktonylee** [microsoft/TypeScript#37700]
## Search Terms

Optional chaining, unknown

## Suggestion

Currently it is forbidden to use optional chaining on `unknown`. But for all possible types in JavaScript, optional chaining will return `undefined` if the field does not exist. So we can safely regard optional chaining of `unknown` type to be `unknown` as well.

This is the verification of optional chaining on all possible types in JavaScript:

```javascript
const num = 5;
const str = '6';
const undef = undefined;
const nul = null;
const bool = true;
const sym = Symbol('sym');
const obj = {};
const func = () => {};

console.log(num?.field);    // print undefined
console.log(str?.field);    // print undefined
console.log(undef?.field);  // print undefined
console.log(nul?.field);    // print undefined
console.log(bool?.field);   // print undefined
console.log(sym?.field);    // print undefined
console.log(obj?.field);    // print undefined
console.log(func?.field);   // print undefined
```

## Use Cases

This is to simplify type checking on nested object with unknown type. For example, we should be able to do something like this:

```typescript
function getFieldFromJson(json: unknown): number {
  const field = json?.fieldA?.fieldB?.fieldC; // <--- this
  if (typeof field === 'number') {
    return field;
  } else {
    return -1;
  }
}
```

Currently we will need to cast explicitly on `someUnknown`, `someUnknown.fieldA`, `someUnknown.fieldA.fieldB` and `someUnknown.fieldA.fieldB.fieldC`. People will find this is unnecessary verbose and may just fallback to use `any`.

## Examples

As above

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by ilogico** [comment]
This is as safe as allowing property access on `{}` (`unknown` is pretty much the same as `{} | null | undefined`)  or allowing access to unknown properties in interfaces in general.
I don't think you should be able to do something with a type and not with its sub-types (we already have `any` for these things).

And if this would be allowed, why should it be restricted to the optional chaining syntax?
```ts
function f(value: unknown): unknown {
  if (value != null) return value.someProp;
  return undefined;
}
```
I believe it is semantically equivalent.

**Comment by RyanCavanaugh** [comment]
I think the real weirdness is that it'd break the operational directionality of `unknown`. Normally for any type `S` that's a subtype of `T`, if an operation is allowed on `T` then it's definitely allowed on `S`, but now `unk?.prop` is valid when e.g. `({ x: 1 })?.prop` isn't even though `{ x: 1 }` is by definition a subtype of `unknown`.

This would lead to some subtle breakage in the future, e.g. let's say you wrote
```ts
function fn(x: unknown) {
  if ('y' in x) {
    console.log(x?.z);
  }
}
```
The proposal is to make this code OK, but then in the future it'd be a breaking change for us to narrow `x` to `{ y: unknown }` in the body of the `if`, which seems bad - narrowing should only ever *increase* the set of things you're allowed to do.

**Comment by falsandtru** [comment]
Should use `x?.['y']` instead.

**Comment by hktonylee** [comment]
> I think the real weirdness is that it'd break the operational directionality of `unknown`. Normally for any type `S` that's a subtype of `T`, if an operation is allowed on `T` then it's definitely allowed on `S`, but now `unk?.prop` is valid when e.g. `({ x: 1 })?.prop` isn't even though `{ x: 1 }` is by definition a subtype of `unknown`.
> 
> This would lead to some subtle breakage in the future, e.g. let's say you wrote
> 
> ```ts
> function fn(x: unknown) {
>   if ('y' in x) {
>     console.log(x?.z);
>   }
> }
> ```
> 
> The proposal is to make this code OK, but then in the future it'd be a breaking change for us to narrow `x` to `{ y: unknown }` in the body of the `if`, which seems bad - narrowing should only ever _increase_ the set of things you're allowed to do.

Hi Ryan thank you for your excellent example. I agree that narrowing should only ever increase the set of things that are allowed to do. But I can think of a counterexample involving the use of contravariance.

```
interface Action {
    actionType: string;
}

interface KeyboardAction {
    actionType: 'keyboard';
    key: string;
}

let genericHandler: (payload: Action) => void;

function isAction(a: unknown): a is Action { return true; }

function isKeyboardAction(a: unknown): a is KeyboardAction { return true; }

function test(unknownAction: unknown) {
    // This will report error for subclass
    if (isKeyboardAction(unknownAction)) {
        type Action = typeof unknownAction;
        const handler: (action: Action) => void = {} as any;
        genericHandler = handler; // <-- Error here
    }

    // Same code as above, but this will *NOT* report error for superclass
    if (isAction(unknownAction)) {
        type Action = typeof unknownAction;
        const handler: (action: Action) => void = {} as any;
        genericHandler = handler;
    }
}
```

As you see class narrowing does not always increase the set of actions that are allowed to do. I believe this `?.` operator also faces similar situation.

Beside theoretical issue, in practice use, people will just be fine to fix the optional chaining of a variable when they decide to narrow the type. Comparing to using `any`, this feature will indeed benefit more than the harms.

Regarding the breaking change concern, if this proposal is published along with the type narrowing of unknown of `in` operator, we should be free from introducing breaking change.

**Comment by omeid** [comment]
> ```ts
> function f(value: unknown): unknown {
>   if (value != null) return value.someProp;
>   return undefined;
> }
> ```



> I think the real weirdness is that it'd break the operational directionality of `unknown`. Normally for any type `S` that's a subtype of `T`, if an operation is allowed on `T` then it's definitely allowed on `S`, but now `unk?.prop` is valid when e.g. `({ x: 1 })?.prop` isn't even though `{ x: 1 }` is by definition a subtype of `unknown`.

Isn't this also true for `any`?




> 
> This would lead to some subtle breakage in the future, e.g. let's say you wrote
> 
> ```ts
> function fn(x: unknown) {
>   if ('y' in x) {
>     console.log(x?.z);
>   }
> }
> ```
> 

I think you meant `console.log(x?.y)`?

> The proposal is to make this code OK, but then in the future it'd be a breaking change for us to narrow `x` to `{ y: unknown }` in the body of the `if`, which seems bad - narrowing should only ever _increase_ the set of things you're allowed to do.

It does increase it from `unknown` to `{ y: T }` which means we can safely do things like `typeof x.y` which wouldn't be possible on `unknown`.

**Comment by omeid** [comment]
For a concrete example, the famous react-router's `Route` component's children prop's location state can not be  _known_ and so it rightly sets the `location` param to `Location<unknown>`.

Now referring to the state is pretty common, here is an example similar to the official documentation:

```typescript
function LoginPage() {
  const location = useLocation();

  const { from } = location.state || { from: "/" } };
  ....
  };

```

Unpacking `from` with typescript when `state` is unknown becomes pretty cumbersome without allowing optional chaning to write:
```typescript
const { from } = typeof state?.from === "string" ? state : { from: "/" }`
```

Or maybe there is a way to handle this simply without optional chaining on unknown?

**Comment by omeid** [comment]
@ilogico 
> ```ts
> function f(value: unknown): unknown {
>   if (value != null) return value.someProp;
>   return undefined;
> }
> ```

Considering that this is possible with `any`, why shouldn't it be possible with `unknown`? At least with `unknown` unlike `any`, it is clear that nothing is known about the return value of this function and you must narrow the type before using it, unlike if it had returned `any`.

**Comment by RyanCavanaugh** [comment]
> Considering that this is possible with any, why shouldn't it be possible with unknown?

Why should two types designed for completely opposite purposes (`any` being the most-permissive, `unknown` the least) behave the same?

**Comment by ljharb** [comment]
@RyanCavanaugh becauae the entire purpose of `?.` is to be able to navigate through all JS types - ie, `x?.foo` is guaranteed to *always* be `unknown`, in JS, when `x` is `unknown`. It may be that TS's design means that that's not a valid property to extract based on the type of the LHS and the name of the property on the RHS - but I'm not sure why that would result in anything but the same result as `x.foo`.

In other words, the only difference between `x.foo` and `x?.foo` should be when `x` can be null or undefined - in which case, `x?.foo` results in `undefined`. As it is now, TS codebases are *forced* to use `any` instead of `unknown` when using optional chaining, which makes the code less type-safe.

**Comment by RyanCavanaugh** [comment]
@ljharb I can't tell what you're trying to say with that.

If you desugared
```ts
function f(value: unknown): unknown {
  return value?.foo;
}
```
into
```ts
function f(value: unknown): unknown {
  return value == null ? undefined : value.foo;
}
```
TypeScript would error on that, same as we would error on `x.foo` if `x` were `object` or `{ }`. Are you saying it should error, or should not error?

It seems like "New syntax constructs should behave like their desugarings" would be an uncontroversial interpretation but that's not what I'm hearing here.

**Comment by ljharb** [comment]
@RyanCavanaugh `value.foo`, specifically, would error, you're right, but let's say it was `value?.toString` - that shouldn't error - it'd either be `undefined` or a toString function.

**Comment by omeid** [comment]
> @RyanCavanaugh 
> > Considering that this is possible with any, why shouldn't it be possible with unknown?
> 
> Why should two types designed for completely opposite purposes (`any` being the most-permissive, `unknown` the least) behave the same?

`any` isn't actually more permissive, it is TypeScript's coercion rules that allows you to use it as any type you like. Both `any` and `unknown` indicate lack of type information.

The reason the parallel is important is because as already mentioned, with `unknown` you can communicate the _unsafe_ nature of your code and afford a degree of type-safety that isn't possible with `any`. So it is important to improve the ergonomics of working with `unknown`. This is consistent with TypeScript design goal of ergonomics over soundness.

**Comment by RyanCavanaugh** [comment]
Nit: Ergonomics isn't necessarily more important than soundness, it's just that it's *allowed to be considered* (in contrast to languages that would *never* make a soundness/ergonomics trade-off).

**Comment by RyanCavanaugh** [comment]
> that shouldn't error - it'd either be undefined or a toString function.

What's the rule that says it's OK for `toString` but not OK for `foo`? Not all non-`null` non-`undefined` JS values have a `toString`, which is equally true of `foo`

**Comment by ljharb** [comment]
@RyanCavanaugh that's true, but since TS is incapable of properly handling things with a [[Prototype]] of `null`, everything in TypeScript that's not `null` or `undefined` does in fact have a `.toString` :-)

**Comment by omeid** [comment]
> @RyanCavanaugh 
> Nit: Ergonomics isn't necessarily more important than soundness, it's just that it's allowed to be considered (in contrast to languages that would never make a soundness/ergonomics trade-off).

Even better, that means since `unknown` allows  _more soundness_ (by reducing the instance of runtime values not matching the inferred type during compile), we should make it possible to use `unknown` in more places which people otherwise would opt for `any`.

**Comment by openreply-dleinhaeuser** [comment]
My information is, that it works like this:

Using `.?` on a value with insufficient type information is what it is for. It does not simply do a check for `null` or `undefined`. It returns `undefined` whenever a property with a name equal to its right hand side argument can not be resolved from its left hand side argument for whatever reason. Be it because such a property does not exist or the RHS value is of a type that does not have properties. The operation is perfectly well defined for values of any type whatsoever.

- If the type of the LHS value is `unknown` or `any` or otherwise has no information about a property of that name (but allows for its possible existence), the result type is `unknown`.
- If the type of the LHS value excludes the existence of such a property, the result type is `undefined`.
- If the type of the LHS value knows the type of a property of that name, the result type is the type of that property (if the property is not be optional, a linter should arguably generate a warning here)

I don't know of any way to make `?.` not work. There is IMHO no change in permissiveness since it is always allowed. It is designed to be always allowed.

**Comment by WIStudent** [comment]
Every time i need to handle an error I am landing back at this issue. Considering that the usage of `unknown` in the catch clause is on the roadmap for 4.3, I think the calls for optional chaining with `unknown` will only increase. Take a look at this example I am currently facing in my code base:

```typescript
const handleError = (error: unknown): void => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const errorType: unknown = (error as any)?.graphQLErrors?.[0]?.errorType;
  if (errorType === 'MyErrorType') {
    handleMyError();
    return;
  }
  throw error;
};

try {
  makeApolloGraphqlClientQuery();
} catch (error) {
  handleError(error);
}
```
The apollo graphql client puts the `errorType` that is returned by the backend deep inside it's own custom error object. Without setting `error` to any in the function definition the only solution I have come up with so far is to cast `error` to any at the start of the optional chain and then casting the result back to `unknown`.

**Comment by bacebu4** [comment]
> Every time i need to handle an error I am landing back at this issue. Considering that the usage of `unknown` in the catch clause is on the roadmap for 4.3, I think the calls for optional chaining with `unknown` will only increase. Take a look at this example I am currently facing in my code base:
> 
> ```ts
> const handleError = (error: unknown): void => {
>   // eslint-disable-next-line @typescript-eslint/no-explicit-any
>   const errorType: unknown = (error as any)?.graphQLErrors?.[0]?.errorType;
>   if (errorType === 'MyErrorType') {
>     handleMyError();
>     return;
>   }
>   throw error;
> };
> 
> try {
>   makeApolloGraphqlClientQuery();
> } catch (error) {
>   handleError(error);
> }
> ```
> 
> The apollo graphql client puts the `errorType` that is returned by the backend deep inside it's own custom error object. Without setting `error` to any in the function definition the only solution I have come up with so far is to cast `error` to any at the start of the optional chain and then casting the result back to `unknown`.

you should be able to rewrite that line like that in order to not disable eslint

`const errorType: unknown = error?.['graphQLErrors']?.[0]?.errorType;`

Thanks @falsandtru for help in figuring this out

**Comment by WIStudent** [comment]
> you should be able to rewrite that line like that in order to not disable eslint
> 
> `const errorType: unknown = error?.['graphQLErrors']?.[0]?.errorType;`

This does not work for me, I get the error "Object is of type 'unknown'.(2571)".
https://www.typescriptlang.org/play?ts=4.3.5#code/MYewdgzgLgBApgJwSBAuGBXMBrMIDuYMAvJmACZwBmAlmHOQNwCwAUKJLALZwQQCGAczgl4SFAH4AdAG0A5Dz5C4cgLotWQA

**Comment by bacebu4** [comment]
> > you should be able to rewrite that line like that in order to not disable eslint
> > `const errorType: unknown = error?.['graphQLErrors']?.[0]?.errorType;`
> 
> This does not work for me, I get the error "Object is of type 'unknown'.(2571)".
> Trying it in playground: https://www.typescriptlang.org/play?ts=4.3.5#code/MYewdgzgLgBApgJwSBAuGBXMBrMIDuYMAvJmACZwBmAlmHOQNwCwAUKJLALZwQQCGAczgl4SFAH4AdAG0A5Dz5C4cgLotWQA

Got it working on Quokka but apparently it won't compile 
Though it have been compiled on my project where deps are over 2 years old. But even after replicating them on freshly created project that is still not working

Sorry for misleading information...

**Comment by hraban** [comment]
> @ljharb I can't tell what you're trying to say with that.
> 
> If you desugared
> 
> ```ts
> function f(value: unknown): unknown {
>   return value?.foo;
> }
> ```
> 
> into
> 
> ```ts
> function f(value: unknown): unknown {
>   return value == null ? undefined : value.foo;
> }
> ```
> 
> TypeScript would error on that, same as we would error on `x.foo` if `x` were `object` or `{ }`. Are you saying it should error, or should not error?
> 
> It seems like "New syntax constructs should behave like their desugarings" would be an uncontroversial interpretation but that's not what I'm hearing here.

Perhaps if we looked at it from a different perspective: `value.foo` isn't the problem; using its value without checking it would be. And that's handled properly by TS, because the value is `unknown`.

Basically, `unknownThing?.foo` guarantees that you get _something_, but you don't know what. That's what the `unknown` type is for, so as long as the resulting expression is also `unknown`, it'd make sense.

Or another perspective: the desugaring is in JS semantics, not TS semantics. We expect `x?.foo` to desugar into `(x == null) ? undefined : x.foo` in JavaScript, but that's because `x.foo` in JS has its own meaning. In TS that would be `(x as any).foo`. So in TS we could say, the desugaring of x.foo is semantically equivalent to (for example) `(x == null) ? undefined : (x as typeof x & { foo: unknown }).foo`. This still fits the JS spec of `?.`, if I understand it correctly?

I don't want to change a spec post-hoc, I'm just trying to explain where I come from re intuition of semantics of this feature. Curious to hear your thoughts.

**Comment by kralphs** [comment]
I mentioned this in a related thread, but I think just coming at this from a set theory/mathematical perspective should carry some weight. No matter how we want to "think" about `unknown`, it represents the entire type space. So if we want to know what the result of optional chaining should be, it should be the range of the optional chaining operator i.e. the result of it acting on all types. If the property exists, the result is still `unknown`, and if it doesn't then it's `undefined`, but `unknown | undefined === unknown`.

Granted, I'm playing a little loose with the math... we'd have to consider the operator acting on an infinite union of sets and whether that's the infinite union of the operator acting on those sets... but I think it's safe to say that when you're dealing with the range being _all_ types, you can at least intuitively conclude that the range as being all types. But hey, if you can find an example of a type that optional chaining _can't_ map to, then you win.

Personally, I feel like the set theoretical interpretation of the type system is foundational. It's really helped me to explain to new devs how `unknown` is the entire space and `never` is the empty set, and draw Venn diagrams to explain why union and intersection types act the way they do.

Likewise, keeping this in mind we can see that using `unknown` is not type narrowing, it's type expanding. That's the whole reason why if you attempt an assertion when two types don't have sufficient overlap, you are encouraged to cast to unknown first because all types can be narrowed down to from `unknown`.

And although it's been mentioned multiple times, I just want to quickly reiterate how this would _improve_ type safety as the compiler would then force us to use optional chaining, as opposed to when `any` is used, and all bets are off.

**Comment by nkappler** [comment]
> And although it's been mentioned multiple times, I just want to quickly reiterate how this would *improve* type safety as the compiler would then force us to use optional chaining, as opposed to when `any` is used, and all bets are off.

I think you can't mention this often enough, as it outweighs all the counter-arguments in my opinion... 😅

**Comment by ESRogs** [comment]
> Every time i need to handle an error I am landing back at this issue. Considering that the usage of unknown in the catch clause is on the roadmap for 4.3, I think the calls for optional chaining with unknown will only increase.

Just chiming in to say that @WIStudent's prediction has come true. I am one of the people who has updated my TypeScript version and now wishes I could use optional chaining in catch clauses.

**Comment by DanielRosenwasser** [comment]
This is another place where "technically works" doesn't mesh well with our philosophy of preventing real errors that you would write.  Allowing `?.` on any value is a big hammer, and we don't think a special case for `unknown`, `object`, and `{}` prevents them enough (nor would it feel good).

One anecdote I'll provide here - a while back, there was a team decision to allow property accesses on any object with a string index signature; after all, any property access (`foo.bar`) is functionally the same as a bracketed element access (`foo["bar"]`). That made a bunch of things easier, but it also introduced bugs - bugs that actually showed up in our own compiler! You can't put that genie back in the lamp without a flag, so now you have [`--noPropertyAccessFromIndexSignature`](https://www.typescriptlang.org/tsconfig#noPropertyAccessFromIndexSignature), an option that [someone who's been on the team for over 7 years](https://github.com/DanielRosenwasser) literally cannot remember the name of.

**Comment by hraban** [comment]
> This is another place where "technically works" doesn't mesh well with our philosophy of preventing real errors that you would write. Allowing ?. on any value is a big hammer, and we don't think a special case for unknown, object, and {} prevents them enough (nor would it feel good).

Thanks for the explanation. Could I ask what exactly you mean? Why are the proposed semantics for propagating `unknown` through `?.` a big / not good hammer? Is it unsound?

Do you know of a real error that people would write with this? I hate to ask for an example of a theoretical problem, but on the other hand, I'm literally writing real errors now, without this feature, by casting to `any` which allows me to skip `?.` checks and type check on the final value. E.g. in JSON decoding.

How do you propose writing a safe JSON decoding routine? E.g., access `{"a":{"b":{"c":123}}}` from an external source, iow including checking every field and final type. Or how does one access deep fields on errors which are `unknown`, without casting to `any` and running the risk of forgetting a `?.`, or forgetting a type check on the last value?

Maybe there's a common idiom I'm missing which makes `?.` propagation on `unknown` less important. Using `any` isn't it, because I'm introducing bugs today. But currently, casting JSON.parse() output to `unknown` is unwieldy.

**Comment by ljharb** [comment]
@DanielRosenwasser so am i understanding properly that then, as well as now, you’re thinking that the migration pain isn’t worth the bugs it will catch, but that it should have been this way from the start?

**Comment by Pimm** [comment]
> How do you propose writing a safe JSON decoding routine? E.g., access `{"a":{"b":{"c":123}}}` from an external source, iow including checking every field and final type. Or how does one access deep fields on errors which are `unknown`, without casting to `any` and running the risk of forgetting a `?.`, or forgetting a type check on the last value?

What I do, is upcast to a conservative version of the type I expect.

In your JSON decoding example, I would do:
```typescript
type Unreliable<T> = { [P in keyof T]?: Unreliable<T[P]> } | undefined;

const decodedData = JSON.parse(data) as Unreliable<{ a: { b: { c: number }}}>;

decodedData?.a?.b?.c; // Type is number | undefined
```
The compiler enforces the use of the `?.` operator ‒ the `.` operator is not allowed, unlike with `any`.

If you would like the compiler to enforce the use of the `typeof` operator as well, you could do:
```typescript
JSON.parse(data) as Unreliable<{ a: { b: { c: unknown }}}>;
```
It's a bunch of extra characters, but comes with the advantage of only allowing access to properties you declared as _potentially_ existing. (No `?.toStirng`.)

**Comment by omeid** [comment]
> This is another place where "technically works" doesn't mesh well with our philosophy of preventing real errors that you would write.

This will allow people to actually use `unknown` where they use `any` now. I do not understand how the philosophy works here. Can you please explain?

**Comment by hraban** [comment]
> Based off the mention in #46314, I would guess that the potential bug would in the class of misspellings of properties i.e. `a?.toStirng()` would almost certainly be an error, but would not cause a run-time error leading to a difficult to find mistake. That's probably the best argument I've seen against it.

Thanks for your reply, I just wanted to focus on the `toStirng` part: aren't typos in type declarations _always_ a problem? If you force type declarations (e.g. @Pimm's proposal) you end up shifting the typo from point of use to point of declaration.

The comment you linked to (#46314) says short-circuiting would create the following risk:

```ts
if (a?.foo) {
    a?.boo.yadda(); // oops
}
```

Maybe someone has a more realistic example, because this is precisely the kind of code you use optional chaining to avoid. You'd write:

```ts
a?.foo.yadda();
```

that's the point of null coalescing member access, as I understand it. But how would you do this currently? Cast to `any`? Same problem. Type casting to a custom type with explicit checking? Same problem, right?

I guess my question is: Is there a solution to handling `unknown` in TS, currently, which is safer than optional chaining? I don't see optional chaining as a concession on any front, I'd love to understand what we'd be giving up.

**Comment by kralphs** [comment]
> Thanks for your reply, I just wanted to focus on the toStirng part: aren't typos in type declarations always a problem? If you force type declarations (e.g. @Pimm's proposal) you end up shifting the typo from point of use to point of declaration.

I was just saying that's the best argument against, not that I was a good argument against lol.

I'm personally advocating for optional chaining to be allowed on unknown for mathematical reasons. Set theory, the basis of the type system itself, and the nature of the `?.` operator seem to indicate optional chaining should be allowed. It both provides justification for its use, and a concrete ground to argue against it being expanded to other operators, thus avoiding the slippery slope that gets bandied about.

The case of typos is a valid concern, but it becomes a question of what's more important: a logically consistent type system or avoiding a class of run-time errors. The type system is supposed to help avoid the latter, but does that make it less important? I don't think so, but others clearly disagree.

But even the argument about avoidance of runtime errors has to be taken as what actually is going to prevent more run time errors, not what might. That ignores the human element. If more of the community stops casting to any because of optional chaining on unknown, that's safer code.

**Comment by Jcparkyn** [comment]
For those interested, you can somewhat emulate this behaviour like this:
```typescript
type SafeAny = {
  [key: PropertyKey]: SafeAny,
} | undefined | null;

const x = null as SafeAny;
console.log(x?.children?.[0]?.age); // This is allowed
console.log(x?.children[0]?.age); // This is an error
```

The caveats here are:
* You need to explicitly cast most types to `SafeAny` (`... as any as SafeAny`)
* This doesn't allow function calls, with or without `?.()`.
* Type narrowing is difficult, because `typeof` doesn't work correctly on this type (technically it is correct, but because we're lying to the compiler it makes the wrong assumptions). You can work around this by using type guards instead.

**Comment by Alexandre-Fernandez** [comment]
this is how I manage this problem (feel free to use it or improve it) :
```ts
export const getUnknownProperty = <T extends Typeof>(
  unknown: unknown,
  propertyKey: PropertyKey,
  typeofProperty: T
): TypeSafePropertyValue<T> | undefined => {
  if (!unknown) return undefined;
  if (typeof unknown !== "object") return undefined;
  const unknownAsRecord = unknown as Record<PropertyKey, unknown>;
  if (typeof unknownAsRecord?.[propertyKey] === typeofProperty) {
    return unknownAsRecord?.[propertyKey] as TypeSafePropertyValue<T>;
  }
  return undefined;
};

type TypeSafePropertyValue<T extends Typeof> = Extract<
  Typeof,
  T
> extends "bigint"
  ? number
  : Extract<Typeof, T> extends "boolean"
  ? boolean
  : Extract<Typeof, T> extends "function"
  ? Function
  : Extract<Typeof, T> extends "number"
  ? number
  : Extract<Typeof, T> extends "object"
  ? object | null
  : Extract<Typeof, T> extends "string"
  ? string
  : Extract<Typeof, T> extends "symbol"
  ? symbol
  : undefined;

type Typeof =
  | "bigint"
  | "boolean"
  | "function"
  | "number"
  | "object"
  | "string"
  | "symbol"
  | "undefined";
```

**Comment by yukinotech** [comment]
> ```ts
> decodedData?.a?.b?.c; // Type is number | undefined
> ```


`decodedData?.a?.b?.c`  should be `unknown` in `JSON.parse()` ,or it is not safe


```ts
const c = decodedData?.a?.b?.c  // Type is number | undefined
if(c){
   c // number , but is it really a number type in js runtime ? 
}

type Unreliable<T> = T extends object
  ? {
      [key in keyof T]-?: Unreliable<T[key]>
    }
  : unknown

```

**Comment by zanminkian** [comment]
```ts
function isTom(cat: unknown) {
  const firstName = cat?.name?.first
  if(typeof firstName === 'string' && firstName === 'Tom') return 'Yes'
  else return 'No'
}
isTom('Jerry') // No
isTom(null) // No
isTom(123) // No
```
Code above will cause a compiling error in TS.
To solve this problem, I have to change `unknown` to `any`, or use type assertion. But I believe we **should not encourage to use `any` and type assertion** in TS.

**Comment by NicholasAntidormi** [comment]
Disallowing optional chaining on `unknown` reduces its utility and forces unnecessary boilerplate. There are many common scenarios where this limitation becomes frustrating. For example:  

```typescript
  try {
    // some code
  } catch (err: unknown) {
    console.log(err?.message); // Not allowed
  }
```

```typescript
  const response: unknown = await getResponse();
  const isOk = response?.status === 'ok'; // Not allowed
  if (isOk) ...
```

```typescript
  const data: unknown = await getData();
  const nestedValue = data?.some?.nested?.value; // Not allowed
  if (isWhatINeed(nestedValue)) ...
```

Is it possible to reconsider this issue?

**Comment by JuanGalilea** [comment]
Also it would solve issues like

```typescript
const someObject = { foo: 'bar' };
const keys = Object.keys(someObject); // string[]

for (const key of keys) {
    // error: No index signature with a parameter of type `string` was found on type `{ foo: string }`
    console.log(someObject?.[key]); 
}
```
being able to optional chain anything into any object would just reduce the painpoints caused by [Object.keys not returning keyof type](https://stackoverflow.com/questions/55012174/why-doesnt-object-keys-return-a-keyof-type-in-typescript) which @RyanCavanaugh seems to have asked and answered in stackoverflow just to make it easier for devs struggling against this to find out why it is the way it is.

Given that 
> Object.keys returns keyof T

is the one that was decided to not be true (which is good), at least a change like this would make the struggle of dealing with it less painful.

**Comment by grundb** [comment]
FWIW, here is a little helper function that just takes any value and performs a sequence of optional chaining property accesses on it (accepts any input and always produces `unknown` result):

```typescript
const optionalChain = (
  obj: unknown,
  path: string[]
): unknown => path.reduce((acc, prop) => (acc as any)?.[prop], obj);
```

**Comment by adrian-gierakowski** [comment]
> FWIW, here is a little helper function that just takes any value and performs a sequence of optional chaining property accesses on it (accepts any input and always produces `unknown` result):
> 
> const optionalChain = (
>   obj: unknown,
>   path: string[]
> ): unknown => path.reduce((acc, prop) => (acc as any)?.[prop], obj);

That's a sledgehammer: you only want result to be unknown if you are chaining on an unknown.

**Comment by grundb** [comment]
> > FWIW, here is a little helper function that just takes any value and performs a sequence of optional chaining property accesses on it (accepts any input and always produces `unknown` result):
> > 
> > const optionalChain = (
> >   obj: unknown,
> >   path: string[]
> > ): unknown => path.reduce((acc, prop) => (acc as any)?.[prop], obj);
> 
> That's a sledgehammer: you only want result to be unknown if you are chaining on an unknown. 

For sure. But it gets the job done for the common case of "I have an `unknown` thing and I want another `unknown` thing"

**Comment by adrian-gierakowski** [comment]
LLMs can do better:

```ts
// 1. Recursive Type to traverse T based on keys in P
type GetPath<T, P extends readonly string[]> = P extends
  [infer Key, ...infer Rest]
  ? Rest extends readonly string[] ? Key extends keyof NonNullable<T>
      // Found valid key: Recurse deeper
      // We add | undefined if the parent was nullable (simulating ?.)
      ?
        | GetPath<NonNullable<T>[Key], Rest>
        | (undefined extends T ? undefined : never)
        | (null extends T ? undefined : never)
      // Invalid key or T is unknown: Return unknown per requirements
    : unknown
  : never
  : T // End of path: return the resulting type

// 2. The Implementation
const optionalChain = <T, const P extends string[]>(
  obj: T,
  path: P,
): GetPath<T, P> => {
  // We cast the result because TS cannot verify the complex recursive reduction at runtime
  return path.reduce((acc, prop) => (acc as any)?.[prop], obj) as GetPath<T, P>
}

type User = {
  createdAt?: Date
  profile: {
    name: string
    age: unknown // Explicit unknown
    address?: {
      street: string
    }
  }
}

const user: User = {
  profile: { name: 'Alice', age: 25, address: { street: 'Liberation Street' } },
  createdAt: new Date(),
}

// 1. Happy Path
export const name: string = optionalChain(user, ['profile', 'name'])
export const createdAt: Date | undefined = optionalChain(user, ['createdAt'])
export const street: string | undefined = optionalChain(user, [
  'profile',
  'address',
  'street',
])

// 2. Path stops early (simulates optional chaining)
export const profile: { name: string; age: unknown } = optionalChain(user, [
  'profile',
])

// 3. Path hits 'unknown' property
// Result: unknown (Because user.profile.age is unknown)
export const age = optionalChain(user, ['profile', 'age'])

// 4. Path continues THROUGH 'unknown'
// Result: unknown (Because we can't infer keys of unknown)
export const invalidDeep = optionalChain(user, ['profile', 'age', 'oops'])

// 5. Path does not exist
// Result: unknown
export const badKey = optionalChain(user, ['profile', 'wrongKey'])

```

the only problem is that you don't get autocomplete for path elements

**Comment by mfulton26** [comment]
https://github.com/sindresorhus/dot-prop can be used to "get, set, or delete a property from a nested object using a dot path [or an array path]" that will return `undefined` for nonexistent paths.

**Comment by glebbash** [comment]
@DanielRosenwasser while I understand the architectural dead-ends and narrowing conflicts that led to this issue being closed, the underlying pain point remains a daily friction for many developers.

I agree that making `unknown?.field` resolve to `unknown` globally could introduce subtle bugs, but the current strictness forces developers into unsafe type-casting (as any) for incredibly standard JavaScript patterns.

## Context

Three quite common scenarios where this issue is a real pain is:

> NOTE: showing only the JavaScript versions. If you could show a good TypeScript solutions for these, then the issue should rightfully stay closed.

1. Error handling

    Quite a lot of libraries set some kind of code on an error that you could use to check error type.

    ```js
    try {
       // ....
    } catch (error) {
       if (error?.code !== Lib.RETRIABLE_ERROR_CODE) {
           throw error;   
       }

       // good to retry the call after a timeout
    }
    ```

2. `JSON.parse` results handling in untyped apis

    ```js
    const dataFromApi = JSON.parse(rawDataFromApi)
    const userName = dataFromApi?.user?.name
    if (typeof userName === "string") {
        // value is good to use
    }
    ```

3. Libraries that accept `unknown` type

   ```js
   const MYLIB_ERROR_TAG = Symbol("MyLibError")

   export function isMyLibErrorType(error) {
       return error?.[MYLIB_ERROR_TAG] === true
   }
   ```

## Possible solutions

1. A restricted `dynamic` or `loose_unknown` type

    A type that behaves like exactly like `unknown`, but permits optional chaining/property access to yield further `dynamic` types. This avoids the total type-check shutdown of `any`.

2. A localized compiler flag or block-level directive: `"allowOptionalChainingOnUnknown": true` or `/* @ts-allow-in-next-block: unknown-optional-chaining */`

    While `tsconfig` flags aren't the best looking solution here (although this might be the simplest option to implement), allowing a block-scoped directive (similar to ESLint rules) to permit optional chaining on `unknown` could localize the risk to just those error-handling or json handling blocks.

There definitely is a middle ground where we don't have to choose between writing verbose (and sometimes runtime affecting) type-guards or completely surrendering to `any` just to safely read a nested property.

I hope this issue can be re-evaluated, I think enough people has raised a voice around this to signal that its often a real pain, especially in codebases that weren't written with TypeScript in mind and are gradually trying to get more type safety without testing every single change after typing existing code, because getting rid of `any`s quite often requires runtime logic changes that are only needed to satisfy the TS compiler.
